File: sdk/python/tokenomics_sdk/wrappers.py
"""Drop-in wrappers for the OpenAI and Anthropic clients.

The wrappers are structural, not subclass-based: they patch the bound ``create`` method
on the client instance. That keeps the original client object -- with all of its typing,
helpers and configuration -- fully intact, and means the SDK does not need the provider
packages installed to import.
"""

from __future__ import annotations

from collections.abc import Callable
from functools import wraps
from typing import Any

from tokenomics_sdk.attributes import Usage, usage_from_anthropic, usage_from_openai
from tokenomics_sdk.tracing import llm_span


def _field(obj: Any, name: str) -> Any:
    return obj.get(name) if isinstance(obj, dict) else getattr(obj, name, None)


def _instrument(
    owner: Any,
    method_name: str,
    *,
    provider: str,
    operation: str,
    extract_usage: Callable[[Any], Usage],
) -> None:
    """Replace ``owner.method_name`` with a span-emitting wrapper (idempotent)."""
    original = getattr(owner, method_name)
    if getattr(original, "__tokenomics_wrapped__", False):
        return

    @wraps(original)
    def wrapper(*args: Any, **kwargs: Any) -> Any:
        request_model = str(kwargs.get("model", "unknown"))
        with llm_span(provider=provider, operation=operation, request_model=request_model) as span:
            response = original(*args, **kwargs)
            usage = _field(response, "usage")
            if usage is not None:
                span.set_usage(extract_usage(usage))
            span.set_response(
                model=_field(response, "model"),
                response_id=_field(response, "id"),
            )
            return response

    wrapper.__tokenomics_wrapped__ = True  # type: ignore[attr-defined]
    setattr(owner, method_name, wrapper)


def wrap_openai[ClientT](client: ClientT) -> ClientT:
    """Instrument an OpenAI client in place and return it."""
    completions = client.chat.completions  # type: ignore[attr-defined]
    _instrument(
        completions,
        "create",
        provider="openai",
        operation="chat",
        extract_usage=usage_from_openai,
    )

    responses = getattr(client, "responses", None)
    if responses is not None:
        _instrument(
            responses,
            "create",
            provider="openai",
            operation="chat",
            extract_usage=usage_from_openai,
        )
    return client


def wrap_anthropic[ClientT](client: ClientT) -> ClientT:
    """Instrument an Anthropic client in place and return it."""
    messages = client.messages  # type: ignore[attr-defined]
    _instrument(
        messages,
        "create",
        provider="anthropic",
        operation="chat",
        extract_usage=usage_from_anthropic,
    )
    return client

File: sdk/typescript/src/wrappers.ts
/**
 * Drop-in wrappers for the OpenAI and Anthropic SDK clients.
 *
 * The wrappers are structural, not subclass-based: they patch the bound `create`
 * method on the client instance. That keeps the original client object -- with all
 * of its typing, helpers and configuration -- fully intact, and means this package
 * never needs `openai` or `@anthropic-ai/sdk` as a dependency.
 *
 * Scope: non-streaming calls only, matching the Python SDK. A streamed response's
 * usage is only known after the last chunk, which needs a different span lifecycle;
 * v0.2 territory, not silently mishandled here (streaming calls fall through to the
 * original method, untraced, rather than emitting a wrong span).
 */

import { usageFromAnthropic, usageFromOpenAI, type Usage } from "./attributes.js";
import { llmSpan } from "./tracing.js";

const WRAPPED = Symbol.for("tokenomics.wrapped");

interface Wrappable {
  (...args: unknown[]): unknown;
  [WRAPPED]?: boolean;
}

function isStreamingRequest(args: unknown[]): boolean {
  const body = args[0];
  return typeof body === "object" && body !== null && (body as { stream?: unknown }).stream === true;
}

function field(obj: unknown, name: string): unknown {
  return obj && typeof obj === "object" ? (obj as Record<string, unknown>)[name] : undefined;
}

function instrument<U>(
  owner: Record<string, unknown>,
  methodName: string,
  options: {
    provider: string;
    operation: string;
    extractUsage: (raw: U) => Usage;
  },
): void {
  const original = owner[methodName] as Wrappable | undefined;
  if (typeof original !== "function" || original[WRAPPED]) return;

  const wrapper: Wrappable = async function (this: unknown, ...args: unknown[]) {
    if (isStreamingRequest(args)) {
      return original.apply(this, args);
    }
    const body = args[0] as Record<string, unknown> | undefined;
    const requestModel = typeof body?.model === "string" ? body.model : "unknown";

    return llmSpan({ provider: options.provider, operation: options.operation, requestModel }, async (span) => {
      const response = await original.apply(this, args);
      const usage = field(response, "usage");
      if (usage !== undefined) {
        span.setUsage(options.extractUsage(usage as U));
      }
      const model = field(response, "model");
      const id = field(response, "id");
      span.setResponse({
        model: typeof model === "string" ? model : undefined,
        responseId: typeof id === "string" ? id : undefined,
      });
      return response;
    });
  };
  wrapper[WRAPPED] = true;
  owner[methodName] = wrapper;
}

// Minimal structural shape of the parts of the OpenAI client this SDK touches.
interface OpenAiLike {
  chat: { completions: Record<string, unknown> };
  responses?: Record<string, unknown>;
}

/** Instrument an OpenAI client in place and return it. */
export function wrapOpenAI<T extends OpenAiLike>(client: T): T {
  instrument(client.chat.completions, "create", {
    provider: "openai",
    operation: "chat",
    extractUsage: usageFromOpenAI,
  });
  if (client.responses) {
    instrument(client.responses, "create", {
      provider: "openai",
      operation: "chat",
      extractUsage: usageFromOpenAI,
    });
  }
  return client;
}

// Minimal structural shape of the parts of the Anthropic client this SDK touches.
interface AnthropicLike {
  messages: Record<string, unknown>;
}

/** Instrument an Anthropic client in place and return it. */
export function wrapAnthropic<T extends AnthropicLike>(client: T): T {
  instrument(client.messages, "create", {
    provider: "anthropic",
    operation: "chat",
    extractUsage: usageFromAnthropic,
  });
  return client;
}
